// api/src/push.ts
import webpush, { type PushSubscription } from "web-push";
import { env } from "./env.js";
import { redis } from "./redis.js";
import type { PushSubscriptionRecord } from "./types.js";

let webPushReady = false;
let webPushDisabled = false;

function initWebPush() {
  if (webPushReady || webPushDisabled) return;

  const pub = env.VAPID_PUBLIC_KEY;
  const priv = env.VAPID_PRIVATE_KEY;

  if (!pub || !priv) {
    webPushDisabled = true;
    console.warn("[web-push] VAPID keys missing; push notifications disabled.");
    return;
  }

  webpush.setVapidDetails(env.VAPID_SUBJECT, pub, priv);
  webPushReady = true;
}

export async function saveSubscription(sub: PushSubscriptionRecord) {
  await redis.sadd("push:subs", JSON.stringify(sub));
}

function shouldRemoveSubscription(err: unknown): boolean {
  const e = err as { statusCode?: number };
  return e?.statusCode === 404 || e?.statusCode === 410;
}

export async function broadcastPush(title: string, body: string, data?: Record<string, unknown>) {
  initWebPush();
  if (webPushDisabled) return;

  const subs = (await redis.smembers("push:subs")) as string[];

  if (!subs.length) return;

  const payload = JSON.stringify({ title, body, data, icon: "/icons/icon-192.png" });

  await Promise.all(
    subs.map(async (s) => {
      try {
        const rec = JSON.parse(s) as PushSubscriptionRecord;

        const sub: PushSubscription = {
          endpoint: rec.endpoint,
          keys: rec.keys,
        };

        await webpush.sendNotification(sub, payload);
      } catch (e) {
        if (shouldRemoveSubscription(e)) {
          await redis.srem("push:subs", s);
        } else {
          console.warn("[web-push] send failed (kept sub):", e);
        }
      }
    })
  );
}
